### PR TITLE
532-refactor: FSD: Widget Course main

### LINKED ISSUE #532: FSD: Widget Course main
## 📝 Describe the refactoring task

Part of the issue #347 

## 📍 Code location

`src/widgets/course-main`

#### `course-main.tsx`
- [x] move `course-main.test.tsx`, `course-main.tsx`, `course-main.module.scss` to `ui` folder
- [x] replace `interface` with `type`
- [x] change `JSX` classes declaration to use `cx` (example in `entities` folder components)
- [x] rename `utils` folder to `helpers`
- [x] move `courseLoader` to `entities/course/api`.
- [x] rename `courseLoader` -> `loader` including file name
- [x] re-export `loader` in the `entities/course/index.ts` file

### PR DESCRIPTION
**Pull request by SkarabeyDM**
## What type of PR is this? (select all that apply)

- [ ] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 🚧 Breaking Change
- [x] 🧑‍💻 Code Refactor
- [ ] 📝 Documentation Update

#### `@/widgets/course-main`

1. move `course-main.test.tsx`, `course-main.tsx`, `course-main.module.scss` to `ui` folder
2. replace `interface` with `type`
3. change `JSX` classes declaration to use `cx` (example in `entities` folder components)
4. rename `utils` folder to `helpers`
5. move `courseLoader` to `entities/course/api`.
6. rename `courseLoader` -> `loader` including file name
7. re-export `loader` in the `entities/course/index.ts` file

## Related Tickets & Documents

- Related Issue #347 #371 
- Closes #532 #596 

## Added/updated tests?

- [ ] 👌 Yes
- [x] 🙅‍♂️ No, because they aren't needed
- [ ] 🙋‍♂️ No, because I need help

## Release Notes

- **New Features**
	- Introduced localization support for course enrollment with English and Russian options.
	- Added a new `HeroCourse` component, replacing the previous `CourseMain` component across various course pages.

- **Improvements**
	- Enhanced `MainTitle` component with additional size variants for better styling options.

- **Refactor**
	- Updated routing to use a more generic loader for course-related pages.

- **Chores**
	- Removed the `CourseMain` component directory and its associated tests, streamlining the codebase.
	- Added comprehensive test coverage for the new `HeroCourse` component and its functionalities.

### CODE REVIEW

File: src/widgets/course-main/ui/course-main.tsx [hunk 1]
```diff
@@ -14,11 +15,13 @@ import { WidgetTitle } from '@/shared/ui/widget-title';
...
+const cx = classNames.bind(styles);
+
+export type CourseMainProps = {
   courseName: string;
   lang?: 'ru' | 'en';
   type?: 'Pre-school RU';
-}
+};
 
 const localizedContent = {
```
Comment (KristiBo) [inline src/widgets/course-main/ui/course-main.tsx:26]: Rename `localizedContent` and move it to the `dev-data` folder

File: src/widgets/course-main/ui/course-main.tsx [hunk 2]
```diff
@@ -1,5 +1,6 @@
+import classNames from 'classnames';
```
Comment (KristiBo) [inline src/widgets/course-main/ui/course-main.tsx:1]:
```suggestion
import classNames from 'classnames/bind';
```

File: src/widgets/course-main/ui/course-main.tsx (hunk 1 above)
Comment (SkarabeyDM) [reply src/widgets/course-main/ui/course-main.tsx:26]: What is the `dev-data` folder? Is it an FSD segment or something else?

Comment (SkarabeyDM) [reply src/widgets/course-main/ui/course-main.tsx:26]: > Rename `localizedContent`

Rename it to what?

Comment (SkarabeyDM) [reply src/widgets/course-main/ui/course-main.tsx]: Done

File: src/widgets/hero-course/ui/hero-course.tsx
Comment (Quiddlee) [inline src/widgets/hero-course/ui/hero-course.tsx:1]: Why did you rename the component from `course-main` to `hero-course`?

Comment (ansivgit) [reply src/widgets/hero-course/ui/hero-course.tsx:1]: https://t.me/c/2142251410/1/6421

File: src/widgets/hero-course/ui/hero-course.module.scss
[the remark this answers is not in the document]
Comment (SpaNb4) [reply src/widgets/hero-course/ui/hero-course.module.scss:1]: @ansivgit @Quiddlee what do you guys think about negative margins? https://github.com/rolling-scopes/site/blob/main/src/widgets/course-main/course-main.module.scss#L18

[the remark this answers is not in the document]
Comment (ansivgit) [reply src/widgets/hero-course/ui/hero-course.module.scss:1]: @SpaNb4 I think it as a crutch and wouldn't like we'd used it in our project.

**Comment by SpaNb4** [comment]
image

**Review by Quiddlee — APPROVED** [review]

File: src/widgets/hero-course/ui/hero-course.test.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+      expect(screen.getByText('available')).toBeVisible();
+    });
+
+    it('renders the section with correct label "UPCOMING"', () => {
+      renderWithRouter(<HeroCourse courseName={angularCourseTitle} />);
+
+      expect(screen.getByText('upcoming')).toBeVisible();
+    });
+  });
```
Comment (SpaNb4) [inline src/widgets/hero-course/ui/hero-course.test.tsx:86]: You don't have to change this, but shouldn't it be tested by adding unit tests for `getCourseStatus` function?

File: src/widgets/hero-course/ui/hero-course.test.tsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  return {
+    ...actual,
+    useLoaderData: vi.fn(() => [mockedCourse, mockedCourseAvailable, mockedCourseUpcoming]),
+  };
+});
+
+const reactCourseTitle = 'React';
+const angularCourseTitle = 'Angular';
+
+const mockedCourse = {
```
Comment (SpaNb4) [inline src/widgets/hero-course/ui/hero-course.test.tsx:20]: `const mockedCourse: Course = {`
and the same for other object

File: src/widgets/hero-course/ui/hero-course.module.scss [hunk 5]
```diff
@@ -0,0 +1,40 @@
+.hero-course {
+  height: 55vh;
```
Comment (SpaNb4) [inline src/widgets/hero-course/ui/hero-course.module.scss:2]: Is this an intentional change?
image

Comment (ansivgit) [reply src/widgets/hero-course/ui/hero-course.module.scss:2]: What is the screenshot on the left? Deploy looks more like the one on the right now. I didn't change the height, I just added `vh`

Comment (SpaNb4) [reply src/widgets/hero-course/ui/hero-course.module.scss:2]: > What is the screenshot on the left? Deploy looks more like the one on the right now. I didn't change the height, I just added `vh`

And that's the reason why it looks different on my monitor

File: src/widgets/hero-course/ui/hero-course.tsx [hunk 6]
```diff
@@ -8,24 +9,22 @@
 import { DateLang } from '@/shared/ui/date-lang';
 import { Image } from '@/shared/ui/image';
 import { LinkCustom } from '@/shared/ui/link-custom';
+import { MainTitle } from '@/shared/ui/main-title';
 import { SectionLabel } from '@/shared/ui/section-label';
 import { Subtitle } from '@/shared/ui/subtitle';
-import { WidgetTitle } from '@/shared/ui/widget-title';
+import { heroCourseLocalized } from 'data';
```
[the remark this answers is not in the document]
Comment (Quiddlee) [reply src/widgets/hero-course/ui/hero-course.tsx:14]: can you remember the fact, that `'data'` is a valid import because of typescript path aliases

File: src/widgets/hero-course/ui/hero-course.test.tsx (hunk 3 above)
Comment (ansivgit) [reply src/widgets/hero-course/ui/hero-course.test.tsx:86]: I didn't change the tests much, grouped them and put them in the general `description`. Adding tests for the function is a good idea

File: src/widgets/hero-course/ui/hero-course.module.scss (hunk 5 above)
Comment (ansivgit) [reply src/widgets/hero-course/ui/hero-course.module.scss:2]: It doesn't look very good. Should I just remove the height in `vh` or is there a better solution?

File: src/widgets/hero-course/ui/hero-course.module.scss [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        font-size: 20px;
+        line-height: 28px;
+      }
+    }
+  }
+
+  @include media-tablet {
+    padding-right: 40px;
+    padding-left: 40px;
+  }
```
[the remark this answers is not in the document]
Comment (ansivgit) [reply src/widgets/hero-course/ui/hero-course.module.scss:39]: This will overwrite the vertical paddings, that's not my goal

File: src/widgets/hero-course/ui/hero-course.test.tsx (hunk 4 above)
Comment (ansivgit) [reply src/widgets/hero-course/ui/hero-course.test.tsx:20]: Actually, it's a part of `Course` type. Should I add to mock all fields or create type for testing with `omit` / `pick`?
image

Comment (SpaNb4) [reply src/widgets/hero-course/ui/hero-course.test.tsx:20]: I'd prefer to have all the properties just in case we use some of them in our logic, e.g. upcoming-courses tests use the following properties
image
but the typeguard requires the use of some additional props, and because of this the behavior in tests and in a real application is different
image

File: src/widgets/hero-course/ui/hero-course.module.scss (hunk 5 above)
Comment (SpaNb4) [reply src/widgets/hero-course/ui/hero-course.module.scss:2]: Yeah, let's just remove it

File: src/widgets/hero-course/ui/hero-course.module.scss [hunk 8]
```diff
@@ -1,37 +1,26 @@
-.container {
+.hero-course {
+  height: 55vh;
```
[the remark this answers is not in the document]
Comment (ansivgit) [reply src/widgets/hero-course/ui/hero-course.module.scss:2]: Done

File: src/shared/ui/main-title/main-title.tsx [hunk 9]
```diff
@@ -8,11 +8,25 @@ type MainTitleProps = HTMLAttributes<HTMLHeadingElement> & VariantProps<typeof m
...
 
-const mainTitleVariants = cva(cx('main-title'));
+const mainTitleVariants = cva(cx('main-title'), {
+  variants: {
+    size: {
+      small: cx('small'),
+      medium: '',
+    },
+  },
+  defaultVariants: { size: 'medium' },
```
[the remark this answers is not in the document]
Comment (ansivgit) [reply src/shared/ui/main-title/main-title.tsx:18]: Fixed

File: src/widgets/hero-course/ui/hero-course.test.tsx (hunk 4 above)
Comment (ansivgit) [reply src/widgets/hero-course/ui/hero-course.test.tsx:20]: Done

File: src/widgets/hero-course/ui/hero-course.module.scss (hunk 5 above)
Comment (ansivgit) [reply src/widgets/hero-course/ui/hero-course.module.scss:2]: Done

File: src/widgets/hero-course/ui/hero-course.test.tsx (hunk 3 above)
Comment (ansivgit) [reply src/widgets/hero-course/ui/hero-course.test.tsx]: Changed
